Parse secrets version 2 data

I can't seem to get vaultenv to work with key-value secrets engine version 2. vaultenv terminates with error 
> [ERROR] Key KEY not found for path PATH

Inspecting Vault's response for version 1 and version 2, it seems likely that the additional 'data' object is causing the problem.

It seems that secrets on vault's key-value/secret engine format 2 is not parsed correctly. (Or documentation is lacking, so I can't understand how to write the secrets file)

In version 2, data is wrapped as such response.data.data.usernmame, here's an actual example from my audit log:


```
    "response": {
        "data": {
            "data": {
                "password": "hmac-sha256:35765f6bd7eb2658b8681e5359c67e5750f13f4f91ecd59210b6dd48a5171dae",
                "username": "hmac-sha256:10f7d662d7052b68c77549181bfa20ee2f26b60d67dd445cad773e23f82fde8d"
            },
            "metadata": {
                "created_time": "hmac-sha256:e37df48fdde45f74021285c9ae71b00dfb367fcf4b93b3bdbd574e9aa20c50a9",
                "deletion_time": "hmac-sha256:32d8ee3d91fe8a45514f70d2d9d0b0d139a494da8dd87ca9405e28d78713acd6",
                "destroyed": false,
                "version": 1
            }
        }
    },
```

vaultenv gives me error-response: 
[ERROR] Key username not found for path data/X/X

my config is
```
VERSION 2

MOUNT prod
report_connection_user=data/X/X#username
```

--- a/app/Main.hs
+++ b/app/Main.hs
@@ -2,9 +2,10 @@
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE OverloadedStrings #-}
 
+import Control.Applicative    ((<|>))
 import Control.Monad          (forM)
-import Control.Lens           (reindexed, to)
 import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.Aeson             (FromJSON, (.:))
 import Data.Bifunctor         (first)
 import Data.List              (nubBy)
 import Data.Monoid            ((<>))
@@ -23,15 +24,13 @@ import Control.Monad.Except   (ExceptT, MonadError, runExceptT, mapExceptT, thro
 import qualified Control.Concurrent.Async   as Async
 import qualified Control.Exception          as Exception
 import qualified Control.Retry              as Retry
-import qualified Data.Aeson.Lens            as Lens (key, members, _String)
+import qualified Data.Aeson                 as Aeson
 import qualified Data.Bifunctor             as Bifunctor
 import qualified Data.ByteString.Char8      as SBS
 import qualified Data.ByteString.Lazy       as LBS hiding (unpack)
 import qualified Data.ByteString.Lazy.Char8 as LBS
 import qualified Data.Foldable              as Foldable
 import qualified Data.Map                   as Map
-import qualified Data.Map.Lens              as Lens (toMapOf)
-import qualified Data.Text                  as Text
 import qualified System.Exit                as Exit
 
 import Config (Options(..), parseOptionsFromEnvAndCli, unMilliSeconds,
@@ -51,7 +50,61 @@ data Context
   , cHttpManager :: Manager
   }
 
-type VaultData = Map.Map String String
+data VaultData = VaultData (Map.Map String String)
+
+-- Vault has two versions of the Key/Value backend. We try to parse both
+-- response formats here and return the one which we can parse correctly.
+--
+-- This means there is some inference going on by vaultenv, but since
+-- we can find out which format we're parsing unambiguously, we just go
+-- with the inference instead of making the user specify this. The benefit
+-- here is that users can upgrade the version of the KV secret engine
+-- without having to change their config files.
+--
+-- KV version 1 looks like this:
+--
+-- @
+-- {
+--   "auth": null,
+--   "data": {
+--     "foo": "bar"
+--   },
+--   "lease_duration": 2764800,
+--   "lease_id": "",
+--   "renewable": false
+-- }
+-- @
+--
+-- And version 2 looks like this:
+--
+-- @
+-- {
+--   "data": {
+--     "data": {
+--       "foo": "bar"
+--     },
+--     "metadata": {
+--       "created_time": "2018-03-22T02:24:06.945319214Z",
+--       "deletion_time": "",
+--       "destroyed": false,
+--       "version": 1
+--     }
+--   },
+-- }
+-- @
+instance FromJSON VaultData where
+  parseJSON =
+    let
+      parseV1 obj = do
+        keyValuePairs <- obj .: "data"
+        VaultData <$> Aeson.parseJSON keyValuePairs
+      parseV2 obj = do
+        nested <- obj .: "data"
+        flip (Aeson.withObject "nested") nested $ \obj' -> do
+          keyValuePairs <- obj' .: "data"
+          VaultData <$> Aeson.parseJSON keyValuePairs
+    in Aeson.withObject "object" $ \obj -> parseV1 obj <|> parseV2 obj
+
 
 -- | Error modes of this program.
 --
@@ -64,6 +117,7 @@ data VaultError
   | KeyNotFound       Secret
   | BadRequest        LBS.ByteString
   | Forbidden
+  | BadJSONResp       String
   | ServerError       LBS.ByteString
   | ServerUnavailable LBS.ByteString
   | ServerUnreachable HttpException
@@ -194,6 +248,7 @@ requestSecret context secretPath =
         ServerUnavailable _ -> True
         ServerUnreachable _ -> True
         Unspecified _ _ -> True
+        BadJSONResp _ -> True
 
         -- Errors where we don't retry
         BadRequest _ -> False
@@ -224,7 +279,7 @@ requestSecrets context secrets = do
 lookupSecrets :: [Secret] -> Map.Map String VaultData -> Either VaultError [EnvVar]
 lookupSecrets secrets vaultData = forM secrets $ \secret ->
   let secretData = Map.lookup (secretRequestPath secret) vaultData
-      secretValue = secretData >>= Map.lookup (sKey secret)
+      secretValue = secretData >>= (\(VaultData vd) -> Map.lookup (sKey secret) vd)
       toEnvVar val = (sVarName secret, val)
   in maybe (Left $ KeyNotFound secret) (Right . toEnvVar) $ secretValue
 
@@ -248,7 +303,7 @@ parseResponse secretPath response =
     responseBody = getResponseBody response
     statusCode = getResponseStatusCode response
   in case statusCode of
-    200 -> Right $ parseSuccessResponse responseBody
+    200 -> parseSuccessResponse responseBody
     403 -> Left Forbidden
     404 -> Left $ SecretNotFound secretPath
     500 -> Left $ ServerError responseBody
@@ -256,12 +311,8 @@ parseResponse secretPath response =
     _   -> Left $ Unspecified statusCode responseBody
 
 
-parseSuccessResponse :: LBS.ByteString -> VaultData
-parseSuccessResponse responseBody =
-  let
-    getter = Lens.key "data" . reindexed Text.unpack Lens.members . Lens._String . to Text.unpack
-  in
-    Lens.toMapOf getter responseBody
+parseSuccessResponse :: LBS.ByteString -> Either VaultError VaultData
+parseSuccessResponse responseBody = first BadJSONResp $ Aeson.eitherDecode' responseBody
 
 --
 -- Utility functions
@@ -284,6 +335,8 @@ vaultErrorLogMessage vaultError =
         "Invalid Vault token"
       InvalidUrl secretPath ->
         "Secret " <> secretPath <> " contains characters that are illegal in URLs"
+      BadJSONResp msg ->
+        "Received bad JSON from Vault: " <> msg
       ServerError resp ->
         "Internal Vault error: " <> (LBS.unpack resp)
       ServerUnavailable resp ->

--- a/app/Main.hs
+++ b/app/Main.hs
@@ -6,9 +6,12 @@ import Control.Applicative    ((<|>))
 import Control.Monad          (forM)
 import Control.Monad.IO.Class (MonadIO, liftIO)
 import Data.Aeson             (FromJSON, (.:))
+import Data.Aeson.Types       (parseMaybe)
 import Data.Bifunctor         (first)
+import Data.HashMap.Strict    (HashMap, lookupDefault, mapMaybe)
 import Data.List              (nubBy)
 import Data.Monoid            ((<>))
+import Data.Text              (Text, pack)
 import Network.Connection     (TLSSettings(..))
 import Network.HTTP.Client    (defaultManagerSettings)
 import Network.HTTP.Conduit   (Manager, newManager, mkManagerSettings)
@@ -19,13 +22,11 @@ import Network.HTTP.Simple    (HttpException(..), Request, Response,
                                getResponseStatusCode)
 import System.Environment     (getEnvironment)
 import System.Posix.Process   (executeFile)
-import Control.Monad.Except   (ExceptT, MonadError, runExceptT, mapExceptT, throwError)
+import Control.Monad.Except   (ExceptT (..), MonadError, runExceptT, mapExceptT, throwError, liftEither, withExceptT)
 
 import qualified Control.Concurrent.Async   as Async
-import qualified Control.Exception          as Exception
 import qualified Control.Retry              as Retry
 import qualified Data.Aeson                 as Aeson
-import qualified Data.Bifunctor             as Bifunctor
 import qualified Data.ByteString.Char8      as SBS
 import qualified Data.ByteString.Lazy       as LBS hiding (unpack)
 import qualified Data.ByteString.Lazy.Char8 as LBS
@@ -38,10 +39,17 @@ import Config (Options(..), parseOptionsFromEnvAndCli, unMilliSeconds,
 import SecretsFile (Secret(..), SFError(..), readSecretList)
 
 -- | Make a HTTP URL path from a secret. This is the path that Vault expects.
-secretRequestPath :: Secret -> String
-secretRequestPath secret = "/v1/" <> sMount secret <> "/" <> sPath secret
+secretRequestPath :: MountInfo -> Secret -> String
+secretRequestPath (MountInfo mountInfo) secret = "/v1/" <> sMount secret <> foo <> sPath secret
+  where
+    foo = case lookupDefault KV1 (pack $ sMount secret <> "/") mountInfo of
+      KV1 -> "/"
+      KV2 -> "/data/"
 
 type EnvVar = (String, String)
+
+data MountInfo = MountInfo (HashMap Text EngineType)
+  deriving (Show)
 
 data Context
   = Context
@@ -49,6 +57,10 @@ data Context
   , cCliOptions :: Options
   , cHttpManager :: Manager
   }
+
+-- | The different types of Engine that Vautlenv supports
+data EngineType = KV1 | KV2
+  deriving (Show)
 
 data VaultData = VaultData (Map.Map String String)
 
@@ -105,6 +117,39 @@ instance FromJSON VaultData where
           VaultData <$> Aeson.parseJSON keyValuePairs
     in Aeson.withObject "object" $ \obj -> parseV1 obj <|> parseV2 obj
 
+-- Parses a very mixed type of response that Vault gives back when you
+-- request /sys/mounts. It has a garbage format. We primarily care about
+-- this information:
+--
+-- {
+--   "secret/": {
+--     "options": {
+--       "version": "1"
+--     },
+--     "type": "kv"
+--   },
+-- }
+--
+-- But there is a whole load of other stuff in the top level object.
+-- Integers, dates, strings, null values. Get the stuff we need
+-- and get out.
+instance FromJSON MountInfo where
+  parseJSON =
+    let
+      getType = Aeson.withObject "MountSpec" $ \o -> do
+        o .: "type" >>= (Aeson.withText "mount type" $ (\case
+          "kv" -> do
+            options <- o .: "options"
+            Aeson.withObject "Options" (\opts -> do
+              version <- opts .: "version"
+              case version of
+                Aeson.String "1" -> pure KV1
+                Aeson.String "2" -> pure KV2
+                _ -> fail "unknown version number") options
+          _ -> fail "expected a KV type"))
+    in
+      Aeson.withObject "MountResp" $ \obj -> do
+        pure . traceShowId $ MountInfo (mapMaybe (\v -> parseMaybe getType v) obj)
 
 -- | Error modes of this program.
 --
@@ -186,8 +231,9 @@ getHttpManager opts = newManager managerSettings
 -- throw HTTP exceptions.
 vaultEnv :: Context -> ExceptT VaultError IO [EnvVar]
 vaultEnv context = do
+  mountInfo <- requestMountInfo context
   secrets <- mapExceptT (fmap $ first SecretFileError) $ readSecretList secretFile
-  secretEnv <- requestSecrets context secrets
+  secretEnv <- requestSecrets context mountInfo secrets
   checkNoDuplicates (buildEnv secretEnv)
     where
       secretFile = oSecretFile (cCliOptions context)
@@ -227,8 +273,24 @@ runCommand options env =
     -- replaces the current process with `command`. It does not return.
     executeFile command searchPath args env'
 
+-- | Look up what mounts are available and what type they have.
+requestMountInfo :: Context -> ExceptT VaultError IO MountInfo
+requestMountInfo context =
+  let
+    cliOptions = cCliOptions context
+    request = setRequestManager (cHttpManager context)
+            $ setRequestHeader "x-vault-token" [SBS.pack (oVaultToken cliOptions)]
+            $ setRequestPath (SBS.pack "/v1/sys/mounts")
+            $ setRequestPort (oVaultPort cliOptions)
+            $ setRequestHost (SBS.pack (oVaultHost cliOptions))
+            $ setRequestSecure (oConnectTls cliOptions)
+            $ defaultRequest
+  in do
+    resp <- withExceptT ServerUnreachable (httpLBS request)
+    withExceptT BadJSONResp (liftEither $ Aeson.eitherDecode' (getResponseBody resp))
+
 -- | Request all the data associated with a secret from the vault.
-requestSecret :: Context -> String -> IO (Either VaultError VaultData)
+requestSecret :: Context -> String -> ExceptT VaultError IO VaultData
 requestSecret context secretPath =
   let
     cliOptions = cCliOptions context
@@ -241,53 +303,70 @@ requestSecret context secretPath =
             $ defaultRequest
 
     -- Only retry on connection related failures
-    shouldRetry _retryStatus (Right _) = return False
-    shouldRetry _retryStatus (Left e) =
-      return $ case e of
-        ServerError _ -> True
-        ServerUnavailable _ -> True
-        ServerUnreachable _ -> True
-        Unspecified _ _ -> True
-        BadJSONResp _ -> True
+    shouldRetry :: Applicative f => Retry.RetryStatus -> VaultError -> f Bool
+    --shouldRetry _retryStatus _ =I--
+    shouldRetry _retryStatus res = pure $ case res of
+      ServerError _ -> True
+      ServerUnavailable _ -> True
+      ServerUnreachable _ -> True
+      Unspecified _ _ -> True
+      BadJSONResp _ -> True
 
-        -- Errors where we don't retry
-        BadRequest _ -> False
-        Forbidden -> False
-        InvalidUrl _ -> False
-        SecretNotFound _ -> False
+      -- Errors where we don't retry
+      BadRequest _ -> False
+      Forbidden -> False
+      InvalidUrl _ -> False
+      SecretNotFound _ -> False
 
-        -- Errors that cannot occur at this point, but we list for
-        -- exhaustiveness checking.
-        KeyNotFound _ -> False
-        DuplicateVar _ -> False
-        SecretFileError _ -> False
+      -- Errors that cannot occur at this point, but we list for
+      -- exhaustiveness checking.
+      KeyNotFound _ -> False
+      DuplicateVar _ -> False
+      SecretFileError _ -> False
 
+    retryAction :: Retry.RetryStatus -> ExceptT VaultError IO VaultData
     retryAction _retryStatus = doRequest secretPath request
   in
-    Retry.retrying (vaultRetryPolicy cliOptions) shouldRetry retryAction
+    retryingExceptT (vaultRetryPolicy cliOptions) shouldRetry retryAction
+
+-- |
+-- Like 'Retry.retrying', but using 'ExceptT' instead of return values. The
+-- predicate also only gets the error, not the return value.
+retryingExceptT
+  :: MonadIO m
+  => Retry.RetryPolicyM m
+  -> (Retry.RetryStatus -> e -> m Bool)
+  -> (Retry.RetryStatus -> ExceptT e m a)
+  -> ExceptT e m a
+retryingExceptT policy predicate action =
+  ExceptT $ Retry.retrying policy predicate' action'
+  where
+    predicate' _ (Right _) = pure False
+    predicate' s (Left e)  = predicate s e
+    action' = runExceptT . action
 
 -- | Request all the supplied secrets from the vault, but just once, even if
 -- multiple keys are specified for a single secret. This is an optimization in
 -- order to avoid unnecessary round trips and DNS requets.
-requestSecrets :: Context -> [Secret] -> (ExceptT VaultError IO) [EnvVar]
-requestSecrets context secrets = do
-  let secretPaths = Foldable.foldMap (\x -> Map.singleton x x) $ fmap secretRequestPath secrets
-  secretDataOrErr <- liftIO $ Async.mapConcurrently (requestSecret context) secretPaths
-  either throwError return $ sequence secretDataOrErr >>= lookupSecrets secrets
+requestSecrets :: Context -> MountInfo -> [Secret] -> ExceptT VaultError IO [EnvVar]
+requestSecrets context mountInfo secrets = do
+  let secretPaths = Foldable.foldMap (\x -> Map.singleton x x) $ fmap (secretRequestPath mountInfo) secrets
+  secretData <- liftIO (Async.mapConcurrently (runExceptT . (requestSecret context)) secretPaths)
+  either throwError return $ sequence secretData >>= lookupSecrets mountInfo secrets
 
 -- | Look for the requested keys in the secret data that has been previously fetched.
-lookupSecrets :: [Secret] -> Map.Map String VaultData -> Either VaultError [EnvVar]
-lookupSecrets secrets vaultData = forM secrets $ \secret ->
-  let secretData = Map.lookup (secretRequestPath secret) vaultData
+lookupSecrets :: MountInfo -> [Secret] -> Map.Map String VaultData -> Either VaultError [EnvVar]
+lookupSecrets mountInfo secrets vaultData = forM secrets $ \secret ->
+  let secretData = Map.lookup (secretRequestPath mountInfo secret) vaultData
       secretValue = secretData >>= (\(VaultData vd) -> Map.lookup (sKey secret) vd)
       toEnvVar val = (sVarName secret, val)
   in maybe (Left $ KeyNotFound secret) (Right . toEnvVar) $ secretValue
 
 -- | Send a request for secrets to the vault and parse the response.
-doRequest :: String -> Request -> IO (Either VaultError VaultData)
+doRequest :: String -> Request -> ExceptT VaultError IO VaultData
 doRequest secretPath request = do
-  respOrEx <- Exception.try . httpLBS $ request :: IO (Either HttpException (Response LBS.ByteString))
-  return $ Bifunctor.first exToErr respOrEx >>= parseResponse secretPath
+  resp <- withExceptT exToErr (httpLBS request)
+  parseResponse secretPath resp
   where
     exToErr :: HttpException -> VaultError
     exToErr e@(HttpExceptionRequest _ _) = ServerUnreachable e
@@ -297,18 +376,18 @@ doRequest secretPath request = do
 -- HTTP response handling
 --
 
-parseResponse :: String -> Response LBS.ByteString -> Either VaultError VaultData
+parseResponse :: (Monad m) => String -> Response LBS.ByteString -> ExceptT VaultError m VaultData
 parseResponse secretPath response =
   let
     responseBody = getResponseBody response
     statusCode = getResponseStatusCode response
   in case statusCode of
-    200 -> parseSuccessResponse responseBody
-    403 -> Left Forbidden
-    404 -> Left $ SecretNotFound secretPath
-    500 -> Left $ ServerError responseBody
-    503 -> Left $ ServerUnavailable responseBody
-    _   -> Left $ Unspecified statusCode responseBody
+    200 -> liftEither (parseSuccessResponse responseBody)
+    403 -> throwError Forbidden
+    404 -> throwError $ SecretNotFound secretPath
+    500 -> throwError $ ServerError responseBody
+    503 -> throwError $ ServerUnavailable responseBody
+    _   -> throwError $ Unspecified statusCode responseBody
 
 
 parseSuccessResponse :: LBS.ByteString -> Either VaultError VaultData

--- a/package.yaml
+++ b/package.yaml
@@ -5,6 +5,7 @@ license: BSD3
 github: channable/vaultenv
 
 dependencies:
+  - aeson
   - base
   - async
   - bytestring
@@ -14,8 +15,6 @@ dependencies:
   - directory
   - http-conduit
   - http-client
-  - lens
-  - lens-aeson
   - megaparsec
   - mtl
   - optparse-applicative

--- a/test/integration_test.sh
+++ b/test/integration_test.sh
@@ -8,8 +8,14 @@ export VAULT_ADDR="http://${VAULT_HOST}:${VAULT_PORT}"
 set -e
 
 # Check the vault command exists:
-if ! which vault; then
+if ! which vault > /dev/null; then
   echo "vault: command not found"
+  exit 1
+fi
+
+if [[ $(vault -version) != "Vault v1"* ]]; then
+  echo "Wrong vault version installed. Integration tests require Vault >= 1.0"
+  echo "Get it here: https://www.vaultproject.io/downloads.html"
   exit 1
 fi
 

--- a/test/integration_test.sh
+++ b/test/integration_test.sh
@@ -51,8 +51,29 @@ testing2#foo
 testing2#bar
 EOF
 
-vault write secret/testing key=testing42 otherkey=testing8 &> /dev/null
-vault write secret/testing2 foo=val1 bar=val2 &> /dev/null
+# First unmount the default `secret/` mount and remount as V1 so we can test
+# with that API version first.
+echo "Running tests for Key/Value mount V1"
+vault secrets disable secret
+vault secrets enable -path=secret -version=1 kv
+
+vault kv put secret/testing key=testing42 otherkey=testing8 &> /dev/null
+vault kv put secret/testing2 foo=val1 bar=val2 &> /dev/null
+
+# Run all tests if we didn't get any arguments.
+if [ $# -eq 0 ]; then
+  echo "Running all tests"
+  echo "NOTE: You might see errors in the log. That's part of what we test."
+  echo "Look at the test summary report to see if there's anything wrong."
+  prove $(find integration -type f -iname '*.sh' ! -name '_*')
+else
+  echo "Running only $1"
+  $1
+fi
+
+# Upgrade the default API.
+echo "Running tests for Key/Value mount V2"
+vault kv enable-versioning secret
 
 # Run all tests if we didn't get any arguments.
 if [ $# -eq 0 ]; then

